### PR TITLE
fix: handled OTHER finish reason gracefully

### PR DESCRIPTION
**Pull request by adubovik**

### CODE REVIEW

**Comment by adubovik** [comment]
/deploy-review
>Deployment status: success

**Comment by adubovik** [comment]
/deploy-review
>Deployment status: success

**Comment by adubovik** [comment]
/deploy-review
>Deployment status: success

**Comment by adubovik** [comment]
/deploy-review
>Deployment status: success

File: aidial_adapter_vertexai/llm/gemini_chat_completion_adapter.py [hunk 1]
```diff
@@ -105,7 +118,36 @@ async def send_message_async(
...
+            ):
+                # Misc finish reason could be usually fixed with a retry
+                raise RetryException(RuntimeError(UNEXPECTED_ERROR_MESSAGE))
+
+            openai_finish_reason = to_openai_finish_reason(finish_reason)
+            if openai_finish_reason is not None:
+                await consumer.set_finish_reason(openai_finish_reason)
+
+            content = get_content(chunk)
+            if content is not None:
```
Comment (Oleksii-Klimov) [inline aidial_adapter_vertexai/llm/gemini_chat_completion_adapter.py:150]: Should an empty string be propagated?

File: aidial_adapter_vertexai/llm/gemini_chat_completion_adapter.py [hunk 2]
```diff
@@ -105,7 +118,36 @@ async def send_message_async(
...
+                log.debug(f"response chunk: {to_dict(chunk)}")
+
+            finish_reason = chunk.candidates[0].finish_reason
+
+            if (
+                completion == ""
+                and finish_reason == Candidate.FinishReason.OTHER
+            ):
+                # Misc finish reason could be usually fixed with a retry
+                raise RetryException(RuntimeError(UNEXPECTED_ERROR_MESSAGE))
```
Comment (Oleksii-Klimov) [inline aidial_adapter_vertexai/llm/gemini_chat_completion_adapter.py]: I think it's worth renaming the exception (e.g. to "class FinishReasonOtherError(RuntimeError)") and avoid nesting another one. The streaming status could be checked in the exception handler.

File: aidial_adapter_vertexai/llm/gemini_chat_completion_adapter.py [hunk 3]
```diff
@@ -105,7 +118,36 @@ async def send_message_async(
...
 
-            yield response.text
+            yield response
+
+    @staticmethod
+    async def process_chunks(
+        consumer: Consumer,
+        generator: Callable[[], AsyncIterator[GenerationResponse]],
+    ) -> AsyncIterator[str]:
+        completion = ""
```
Comment (Oleksii-Klimov) [inline aidial_adapter_vertexai/llm/gemini_chat_completion_adapter.py]: Could be int.

File: aidial_adapter_vertexai/llm/consumer.py [hunk 4]
```diff
@@ -44,31 +50,8 @@ async def add_attachment(self, attachment: Attachment):
...
-    async def add_attachment(self, attachment: Attachment):
-        self.attachments.append(attachment)
-
-    async def set_usage(self, usage: TokenUsage):
-        self.usage = usage
+    async def set_finish_reason(self, finish_reason: FinishReason):
+        if self.finish_reason is None:
+            self.finish_reason = finish_reason
+        else:
+            assert self.finish_reason == finish_reason, "finish reason mismatch"
```
Comment (Oleksii-Klimov) [inline aidial_adapter_vertexai/llm/consumer.py]: The error message in the CollectConsumer is slightly clearer.

File: aidial_adapter_vertexai/llm/gemini_chat_completion_adapter.py (hunk 3 above)
Comment (adubovik) [reply aidial_adapter_vertexai/llm/gemini_chat_completion_adapter.py]: Fixed. Turned into a boolean.

File: aidial_adapter_vertexai/llm/consumer.py (hunk 4 above)
Comment (adubovik) [reply aidial_adapter_vertexai/llm/consumer.py]: Fixed

File: aidial_adapter_vertexai/llm/gemini_chat_completion_adapter.py (hunk 1 above)
Comment (adubovik) [reply aidial_adapter_vertexai/llm/gemini_chat_completion_adapter.py:150]: The chunk didn't have any content, why should we turn it into an empty string?

File: aidial_adapter_vertexai/llm/gemini_chat_completion_adapter.py (hunk 2 above)
Comment (adubovik) [reply aidial_adapter_vertexai/llm/gemini_chat_completion_adapter.py]: Fixed

**Comment by adubovik** [comment]
/deploy-review
>Deployment status: success

File: aidial_adapter_vertexai/llm/gemini_chat_completion_adapter.py (hunk 1 above)
Comment (Oleksii-Klimov) [reply aidial_adapter_vertexai/llm/gemini_chat_completion_adapter.py:150]: I thought an empty string shouldn't be propagated but maybe I was wrong.

**Review by Oleksii-Klimov — APPROVED** [review]

**Pull request merged**